### PR TITLE
update contributing guide

### PR DESCRIPTION
**Pull request by haifeng-jin**
The CLA part is following the [TensorFlow contributing guide](https://github.com/tensorflow/tensorflow/blob/master/CONTRIBUTING.md).

The original contents of the file is largely preserved.
They are merged into different parts of the guide.

A refined version of the "Run tests" section and "Code style guide" section will come later in another pull request.

### CODE REVIEW

**Comment by fchollet** [comment]
> hesitate to suggest user to use other tools like VScode since everyone has their preference for IDEs, same for github codespace. I think we should suggest the option that has the least dependency with other toolsets, and provide docs for alternatives like VS code if user already uses them.

Strong agree with this. It's good to provide people's with an easy way to use their favorite development tools, but we should not prescribe one specific workflow.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -1,51 +1,99 @@
...
-possibly guide you. Coordinating up front makes it much easier to avoid
-frustration later on.
-
-### Code reviews
-
-All submissions, including submissions by project members, require review. We
-use Github Pull Requests (PRs) for this purpose. We recommend you read [this guide](
-https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/about-pull-requests)
-if you haven't created a Pull Request before.
+Before we can use your code, please fill out either the individual or corporate Contributor License Agreement (CLA).
```
Comment (fchollet) [inline CONTRIBUTING.md:3]: Let's mention that the CLA will pop up when you create a PR. You don't need to worry about it before you get started. In fact, let's move the CLA section to *after* the "open a PR" section.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -1,51 +1,99 @@
...
-one above, see the
-[Software Grant and Corporate Contributor License Agreement](https://cla.developers.google.com/about/google-corporate).
+To make code changes, you need to fork the repository.
+You may need to setup a dev environment, run the unit tests, which are introduced in the later sections of this guide.
 
-### Tools needed for development
+#### Step 3. Create a pull request 
+
+Once the code changes are made,
+you can open a pull request from your branch in your fork to the master branch in [keras-team/keras](https://github.com/keras-team/keras)..
```
Comment (fchollet) [inline CONTRIBUTING.md:30]: Keep lines short everywhere (but keep links on a single line).

Remove the extra period at the end.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -1,51 +1,99 @@
...
+
+This option is for advanced contributors.
+
+To setup your local dev environment, you will need the following tools.
 
 1.  [Bazel](https://bazel.build/) is the tool to build and test Keras. See the
     [installation guide](https://docs.bazel.build/versions/4.0.0/install.html)
     for how to install and config bazel for your local environment.
 2.  [git](https://github.com/) for code repository management.
 3.  [python](https://www.python.org/) to build and code in Keras.
```
Comment (fchollet) [inline CONTRIBUTING.md:95]: 3.7 in this case.

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -1,51 +1,99 @@
...
+#### Option 1: Visual Studio Code
+
+This is an easy option for beginners.
+
+Clone your fork of the repo to your local machine with the following command.
+```shell
+git clone https://github.com/YOUR_USERNAME/keras.git
+```
+Open [Visual Studio Code](https://code.visualstudio.com/).
+Install the [Remote-Containers extension](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers).
```
[the remark this answers is not in the document]
Comment (haifeng-jin) [reply CONTRIBUTING.md:64]: Yes, this will use the dockerfile.

**Comment by haifeng-jin** [comment]
@fchollet @qlzh727 
I have removed the VS code and GitHub Codespaces option,
and added more descriptions to the Docker part.
Also addressed the rest of the comments.

I think it is also important to teach the beginner contributors about how to use their favorite IDE to develop Keras.
As suggested by Scott, we can have a link to a separate doc, which is the setup steps for a list of popular IDEs.

Many of the contributors may be a one-time contributor trying to make minor changes with minimal setup steps.
They may find a good setup option in that list.

What do you think?

**Review by fchollet — APPROVED** [review]
LGTM, thank you

**Review by qlzh727 — APPROVED** [review]

**Comment by fchollet** [comment]
> This branch cannot be rebased due to conflicts

Please rebase the commits.

**Pull request merged**